src/neovim.rs

```diff
@@ -180,39 +190,35 @@ where
     }
   }
 
-  async fn io_loop<H, R>(
+  async fn handler_loop<H>(
+    self,
     handler: H,
-    mut reader: R,
-    neovim: Neovim<H::Writer>,
+    mut receiver: UnboundedReceiver<RpcMessage>,
   ) -> Result<(), Box<LoopError>>
   where
-    H: Handler + Spawner,
-    R: AsyncRead + Send + Unpin + 'static,
+    H: Handler<Writer = W> + Spawner,
   {
-    let mut rest: Vec<u8> = vec![];
-
     loop {
-      let msg = match model::decode(&mut reader, &mut rest).await {
-        Ok(msg) => msg,
-        Err(err) => {
-          let e = neovim.send_error_to_callers(&neovim.queue, *err).await?;
-          return Err(Box::new(LoopError::DecodeError(e, None)));
-        }
+      let msg = match receiver.next().await {
+        Some(msg) => msg,
+        None => break Ok(()),
```

Comment:
`.unwrap()` sounds fine, I don't know if this error is actually worth propagating though. When I first was writing this I thought it might be back in https://github.com/KillTheMule/nvim-rs/pull/26#issuecomment-873438358 , but I realized I might be overthinking it - the only reason why `receiver` would close here is because `io_loop()` exited

Reply:
Actually, I think `Ok(())` would be the right thing to do here? `receiver` is only receiving messages from `io_loop()` and not `handler_loop()`. And also, I would think that `receiver` is always going to close at some point anyway once `io_loop()` shuts down - properly or improperly. So I don't think we want to panic here.

Reply:
Ohhh right that would be the normal shutdown procedure - neovim closes stdin/stdout, the io-loop shuts down (and now you mention it, that should not be an error as well, but just `Ok(())`) and takes the receiver with it. `Ok(())` it is, thanks! Please add a little comment to the function for my future self ;)